cfep-09.md

```diff
@@ -0,0 +1,99 @@
+
+<table>
+<tr><td> Pinning Proposal System for Automatic Rebuilds</td><td> Pinning Proposal </td>
+<tr><td> Status </td><td> Proposed </td></tr>
+<tr><td> Author(s) </td><td> Christopher J. "CJ" Wright &lt;cjwright4242@gmail.com&gt;; Matthew R. Becker &lt;becker.mr@gmail.com&gt;</td></tr>
+<tr><td> Created </td><td> Feb 20, 2019</td></tr>
+<tr><td> Updated </td><td> Aug 30, 2016</td></tr>
+<tr><td> Discussion </td><td> NA </td></tr>
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Abstract
+
+This CFEP proposes a change to how our internal pinning system works.
+The new system will comprise of a new repo for proposed migrations.
+To propose a migration a PR will be issued with a new version of
+the global pinning file and some metadata about the migration needed 
+(if any).
+Once the PR is squashed and merged the bot will take the metadata
+to calculate which packages need to be rebuilt and start issuing
+PRs with the new pinning as a local pinning file to the feedstocks
+in topo order.
+
+This CFEP is taken from @beckermr's work on [conda-forge.github.io](https://github.com/conda-forge/conda-forge.github.io/issues/712).
+
+## Motivation
+Currently when we move pinnings we move the pin in the ``conda-forge-pinning-feedstock``.
+Then we start a migration using the ``regro-cf-autotick-bot``.
+This order of operations causes some issues, when new feedstocks
+are created they recieve the new pinnings even if their dependencies
+are not rebuilt with the new pinnings themselves.
+
+## Specification
+
+### The migrations proceed as follows:
+1. We make a PR on the `migrations` repo with a change to the global pinnings file, generating a new version. 
+2. Once the change is merged, the bot picks it up and starts going through the graph in topo order. 
+3. A migration PR is issued for a node only if 
+    1. The node depends on the changed pinnings.
+    2. It has no dependencies that depend on the new pinnings and have not been migrated.
+    3. It is not currently being migrated by an earlier migration. Per @scopatz, this state is stored in `libcflib` with a REST API.
+4. Process 3 continues until we determine that the migration is sufficiently complete and the change is merged into the global pinnings file. 
```

Comment:
I think we need some discussion here about how "sufficiently complete" is stated. We also might need some discussion about what to do with packages that haven't been migrated but the migration is deemed complete.

Reply:
> packages that haven't been migrated but the migration is deemed complete.

This has been an annoying thorn in the side of the compiler migration

Reply:
Right. So my feeling from watching the great compiler migration is that there is a long tail of stuff that just doesn’t get migrated. 

I think we will need settings to tell the bot how critical things are. For critical migrations we could require 100%. For minor version bumps we could be more lax.

Reply:
I think we need a combination of percentage and must have packages.

Reply:
Yep this makes sense.

Reply:
I added a note on this.